File: R/plot_scripts.R
# Plot scripts

t_curv = function (TOST_res,
                   steps = 5000) {
  intrvls <- (0:steps)/steps

  # Confidence interval of the SMD from Goulet-Pelletier & Cousineau
  results <- suppressWarnings({lapply(intrvls, FUN = function(i) t_CI(TOST_res, 1-i))})

  df <- data.frame(do.call(rbind, results))
  intrvl.limit <- c("lower.limit", "upper.limit")
  colnames(df) <- intrvl.limit
  df$intrvl.width <- (abs((df$upper.limit) - (df$lower.limit)))
  df$intrvl.level <- intrvls
  df$cdf <- (abs(df$intrvl.level/2)) + 0.5
  df$pvalue <- 1 - intrvls
  df$svalue <- -log2(df$pvalue)
  df <- head(df, -1)
  class(df) <- c("data.frame", "concurve")
  densdf <- data.frame(c(df$lower.limit, df$upper.limit))
  colnames(densdf) <- "x"
  densdf <- head(densdf, -1)
  class(densdf) <- c("data.frame", "concurve")

  return(list(df, densdf))

}

d_curv = function (TOST_res,
                   steps = 5000) {
  intrvls <- (0:steps)/steps

  # Confidence interval of the SMD from Goulet-Pelletier & Cousineau
  results <-
    suppressWarnings({
      lapply(
        intrvls,
        FUN = function(i)
          d_CI(
            d = TOST_res$smd$d,
            df = TOST_res$smd$d_df,
            lambda = TOST_res$smd$d_lambda,
            sigma = TOST_res$smd$d_sigma,
            t_stat = TOST_res$smd$t_stat,
            #hn = TOST_res$smd$hn,
            smd_ci = TOST_res$smd$smd_ci,
            alpha = 1 - i
          )
      )
    })

  df <- data.frame(do.call(rbind, results))
  intrvl.limit <- c("lower.limit", "upper.limit")
  colnames(df) <- intrvl.limit
  df$intrvl.width <- (abs((df$upper.limit) - (df$lower.limit)))
  df$intrvl.level <- intrvls
  df$cdf <- (abs(df$intrvl.level/2)) + 0.5
  df$pvalue <- 1 - intrvls
  df$svalue <- -log2(df$pvalue)
  df <- head(df, -1)
  class(df) <- c("data.frame", "concurve")
  densdf <- data.frame(c(df$lower.limit, df$upper.limit))
  colnames(densdf) <- "x"
  densdf <- head(densdf, -1)
  class(densdf) <- c("data.frame", "concurve")

  return(list(df, densdf))

}

d_curv_raw = function (d,
                       df,
                       lambda,
                       sigma,
                       smd_ci = "goulet",
                       steps = 5000) {
  intrvls <- (0:steps)/steps

  # Confidence interval of the SMD from Goulet-Pelletier & Cousineau
  results <- suppressWarnings({lapply(intrvls, FUN = function(i) d_CI(d = d,
                                                                      df = df,
                                                                      lambda = lambda,
                                                                      sigma = sigma,
                                                                      smd_ci = smd_ci,
                                                                      t_stat= NULL,
                                                                      #hn = NULL,
                                                                      alpha = 1-i))})

  df <- data.frame(do.call(rbind, results))
  intrvl.limit <- c("lower.limit", "upper.limit")
  colnames(df) <- intrvl.limit
  df$intrvl.width <- (abs((df$upper.limit) - (df$lower.limit)))
  df$intrvl.level <- intrvls
  df$cdf <- (abs(df$intrvl.level/2)) + 0.5
  df$pvalue <- 1 - intrvls
  df$svalue <- -log2(df$pvalue)
  df <- head(df, -1)
  class(df) <- c("data.frame", "concurve")
  densdf <- data.frame(c(df$lower.limit, df$upper.limit))
  colnames(densdf) <- "x"
  densdf <- head(densdf, -1)
  class(densdf) <- c("data.frame", "concurve")

  return(list(df, densdf))

}

File: pyTOSTER/plot_scripts.py
import numpy as np
import pandas as pd
from scipy import stats

def _t_ci(tost_res, alpha):
    est = tost_res['effsize']['estimate'][0]
    se = tost_res['effsize']['SE'][0]
    df = tost_res['TOST']['df'][0]
    t_val = stats.t.ppf(1 - alpha / 2, df)
    return est - t_val * se, est + t_val * se

def _d_ci(d, df, ncp, sigma, smd_ci, t_stat, alpha):
    # Simplified version of d_CI
    if sigma is np.nan:
        return np.nan, np.nan
    return d - 1.96 * sigma, d + 1.96 * sigma

def t_curv(tost_res, steps=5000):
    intrvls = np.linspace(0, 1, steps + 1)
    intrvls = intrvls[(intrvls > 0) & (intrvls < 1)]

    results = [_t_ci(tost_res, 1-i) for i in intrvls]

    df = pd.DataFrame(results, columns=['lower.limit', 'upper.limit'])
    df['intrvl.level'] = intrvls
    df['pvalue'] = 1 - intrvls

    densdf = pd.DataFrame({'x': np.concatenate([df['lower.limit'], df['upper.limit']])})

    return [df, densdf]

def d_curv(tost_res, steps=5000):
    intrvls = np.linspace(0, 1, steps + 1)
    intrvls = intrvls[(intrvls > 0) & (intrvls < 1)]

    smd = tost_res['smd']
    results = [_d_ci(smd['d'], smd.get('d_df'), smd.get('d_lambda'), smd.get('d_sigma'), smd.get('smd_ci'), smd.get('t_stat'), 1-i) for i in intrvls]

    df = pd.DataFrame(results, columns=['lower.limit', 'upper.limit'])
    df['intrvl.level'] = intrvls
    df['pvalue'] = 1 - intrvls

    densdf = pd.DataFrame({'x': np.concatenate([df['lower.limit'], df['upper.limit']])})

    return [df, densdf]

def d_curv_raw(d, df, ncp, sigma, smd_ci="goulet", steps=5000):
    intrvls = np.linspace(0, 1, steps + 1)
    intrvls = intrvls[(intrvls > 0) & (intrvls < 1)]

    results = [_d_ci(d, df, ncp, sigma, smd_ci, None, 1-i) for i in intrvls]

    df = pd.DataFrame(results, columns=['lower.limit', 'upper.limit'])
    df['intrvl.level'] = intrvls
    df['pvalue'] = 1 - intrvls

    densdf = pd.DataFrame({'x': np.concatenate([df['lower.limit'], df['upper.limit']])})

    return [df, densdf]
